[libc++] Implement adjacent_view

huixie90 opened the pull request.

ldionne commented on libcxx/include/__ranges/adjacent_view.h:
```diff
@@ -0,0 +1,408 @@
+    requires(!__simple_view<_View>)
+  {
+    if constexpr (common_range<_View>) {
+      return __iterator<false>(__as_sentinel{}, ranges::begin(__base_), ranges::end(__base_));
```
Under the hood, you basically increment `begin(base)` until `end(base) - size-of-your-window`, right? I assume:
- for random-access, we do `begin(base) + M`
- for bidirectional, we walk back from `end(base)`
- for forward, we walk from `begin(base)` to `end(base) - M`?

ldionne commented on libcxx/include/__ranges/adjacent_view.h:
```diff
@@ -0,0 +1,408 @@
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto size()
```
Could we use deducing-this here to avoid the duplication?

ldionne commented on libcxx/include/__ranges/zip_utils.h:
```diff
@@ -0,0 +1,49 @@
+
```
```suggestion
```

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.adjacent/adaptor.pass.cpp:
```diff
@@ -0,0 +1,242 @@
+  test_constraints<5>();
+}
+
+constexpr void test_zero_case() {
```
```suggestion
static constexpr void test_zero_case() {
```

Suggestion: this will make the compiler issue a warning if you forget to call the function anywhere in the file.

huixie90 commented on libcxx/include/__ranges/adjacent_view.h:
```diff
@@ -0,0 +1,408 @@
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator(iterator_t<_Base> __first, sentinel_t<_Base> __last) {
+    __current_[0] = __first;
+    for (int __i = 1; __i < static_cast<int>(_Np); ++__i) {
+      __current_[__i] = ranges::next(__current_[__i - 1], 1, __last);
```
indeed the codegen is different (I lost my godbolt now :(  )  i need to do a bit benchmark and see.

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.adjacent/helpers.h:
```diff
@@ -0,0 +1,43 @@
+template <std::size_t N, class T>
+using expectedTupleType = typename ExpectedTupleType<N, T>::type;
+
+#endif // TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ADJACENT_HELPERS_H
```
Missing newline at the end.

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.adjacent/iterator/singular.pass.cpp:
```diff
@@ -0,0 +1,106 @@
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+// UNSUPPORTED: no-exceptions
+
+// If the invocation of any non-const member function of `iterator` exits via an
```
If there's a section in the standard that says this, can you add the ref or quote the paragraph?

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.adjacent/iterator/singular.pass.cpp:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  {
+    // adjacent_view iterator should be able to be assigned after member function throws
```
```suggestion
    // adjacent_view iterator should be assignable after member function throws
```

ldionne reviewed (approved):
LGTM, thanks! Just missing the adjustment with the LWG issue number.

huixie90 committed:
rebase

[libc++] implement adjacent_view

ci

ci

ci

ci

Update libcxx/include/__ranges/zip_utils.h

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

review

review

benchmark

format

fix comma

fix comma

lwg4482

rebase error

[LLVM][CMake] Build examples for llvm-test-depends

Build examples and example plug-ins by default when running tests. If
examples are unwanted, they can still be disabled completely using
LLVM_INCLUDE_EXAMPLES=OFF. Plugin tests depend on examples and it is
beneficial to test them by default. By default, Examples will still not
be included in the default target or be installed, this remains
controlled by LLVM_BUILD_EXAMPLES (which defaults to OFF).

The additional cost for building examples for tests is 17 compilation
units (12 C++, 5 C), which should be tolerable.

I don't know how broken the examples currently are in the various build
configurations, but if we find breakage, it would be good to fix it.

Pull Request: https://github.com/llvm/llvm-project/pull/171998

Update libcxx/include/__ranges/adjacent_view.h

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

?

undo rebase

ci
